syntax = "proto3";
import "parameter.proto";
import "google/protobuf/struct.proto";

package index;

/**
    Service to build index
*/
service Indexer {
    rpc GetIndexers(GetIndexersRequest) returns (GetIndexersResponse);
    rpc GenerateIndex(stream GenerateIndexRequest) returns (GenerateIndexResponse);
    rpc GetIndexStatus(GetIndexStatusRequest) returns (IndexStatusResponse);
    rpc GetIndexes(GetIndexesRequest) returns (GetIndexesResponse);
}

message IndexerComponent {
    string indexer_id = 1;
    repeated parameter.Parameter parameters = 2;
}

message GetIndexersRequest {
}

message GetIndexersResponse {
    repeated IndexerComponent indexers = 1;
}

message Document {
    string text = 1;
    string document_id = 2;
    string title = 3;
}

message GenerateIndexRequest {
    IndexerComponent indexer = 1;
    repeated Document documents = 2;
    string index_id = 3;
    google.protobuf.Struct metadata = 4;
}

enum IndexStatus {
    READY = 0;
    INDEXING = 1;
    DOES_NOT_EXISTS = 2;
    CORRUPT = 3;
}

message GenerateIndexResponse {
    string index_id = 1;
    IndexStatus status = 2;
}

message GetIndexStatusRequest {
    string index_id = 1;
}

message IndexStatusResponse {
    IndexStatus status = 1;
}

message GetIndexesRequest {
    string engine_type = 1;
}

message IndexInformation {
    string index_id = 1;
    IndexStatus status = 2;
    google.protobuf.Struct metadata = 3;
}

message GetIndexesResponse {
   repeated IndexInformation indexes = 1;
}